[RFC] vim-patch:8.0.{755,1146,1164}

janlazo opened the pull request:
~~I omitted changes for gui code. Seems applicable for termguicolors but I don't know what to do.~~

8.0.0755 was marked N/A but it has a relevant `syntax.c` change to match the other `GUI*` color checks.

janlazo committed:
vim-patch:8.0.1164: changing StatusLine highlight does not always work

Problem:    Changing StatusLine highlight while evaluating 'statusline' may
            not change the status line color.
Solution:   When changing highlighting while redrawing don't cause another
            redraw. (suggested by Ozaki Kiichi, closes vim/vim#2171, closes vim/vim#2120)
https://github.com/vim/vim/commit/65ed136844fbaffdd473903ed841c944600234dc

justinmk commented:
I think we need 8.0.1146 https://github.com/vim/vim/commit/452030e530aad9b08fcfa71737d098b33c752b85 , no?

janlazo commented:
@justinmk Yes if it's relevant to termguicolors.

erw7 commented:
@janlazo Regardless of `termguicolors`, isn't it a problem with redrawing? Therefore, I think your changes are correct as shown below. I do not think that further work is necessary for this PR.

without this PR:
![without](https://user-images.githubusercontent.com/11682285/60565311-0c674980-9d9e-11e9-8247-bfd72a609148.png)

with this PR:
![with](https://user-images.githubusercontent.com/11682285/60565327-19843880-9d9e-11e9-90bf-b4f9cc2771fb.png)

janlazo commented:
@erw7 Thanks for verifying that the patch works but I want to fully port this patch (and any related patches) in this PR.

janlazo committed:
vim-patch:8.0.0755: terminal window does not have colors in the GUI

Problem:    Terminal window does not have colors in the GUI.
Solution:   Lookup the GUI color.
https://github.com/vim/vim/commit/26af85d97ba1ed0ade6cdd41890ca04ed879b9c7

janlazo commented:
Actually, is 8.0.1146 N/A? I don't understand what the code changes other than call `vim_free` when the condition, `*namep == NULL || STRCMP(*namep, arg) != 0`, is met.

janlazo committed:
vim-patch:8.0.1146: redraw when highlight is set with same names

Problem:    Redraw when highlight is set with same names. (Ozaki Kiichi)
Solution:   Only free and save a name when it changed. (closes vim/vim#2120)
https://github.com/vim/vim/commit/452030e530aad9b08fcfa71737d098b33c752b85

janlazo commented:
Am I missing a patch? The travis failure is because of gcov.

blueyed commented:
> The travis failure is because of gcov.

Restarted, and added to https://github.com/neovim/neovim/issues/10332#issuecomment-509298259 for later investigation.  Currently it baffles me that `$GCOV_ERROR_FILE` is not honored.

blueyed commented:
> Wouldn't a macro be easier for 8.0.1146?

Maybe via Vim, but I think it is better to not factor it out only here.  Assuming you mean the 3 similar code blocks.
